File: Training/TCS_F.py
def room_optimization(room_sizes, initial_land, additional_land):
    def max_rooms(area):
        dp = [0]*(area+1)
        for i in range(1, area+1):
            for room in room_sizes:
                if room <= i:
                    dp[i] = max(dp[i], 1 + dp[i-room])
        return dp[-1]

    initial_rooms = max_rooms(initial_land)
    total_rooms = max_rooms(initial_land + additional_land)
    return total_rooms - initial_rooms

# Take the input for room sizes
A, B, C, D = map(int, input().split())
room_sizes = [A, B, C, D]

# Take the input for initial and additional land
X, Y = map(int, input().split())

# Calculate the difference in the number of rooms
difference = room_optimization(room_sizes, X, Y)

# Print the difference
print(difference)

File: Training/TCS_F.cpp
#include <iostream>
#include <vector>

using namespace std;

vector<int> room_optimization(int A, int B, int C, int D, int X, int Y) {
    vector<int> room_types = {A, B, C, D};
    sort(room_types.begin(), room_types.end(), greater<int>());

    vector<int> initial_rooms(4, 0);
    vector<int> final_rooms(4, 0);
    int total_area = X + Y;

    // Calculate initial rooms
    for (int i = 0; i < 4; ++i) {
        while (X >= room_types[i]) {
            X -= room_types[i];
            initial_rooms[i]++;
        }
    }

    // Calculate final rooms
    for (int i = 0; i < 4; ++i) {
        while (total_area >= room_types[i]) {
            total_area -= room_types[i];
            final_rooms[i]++;
        }
    }

    // Calculate difference
    vector<int> difference(4, 0);
    for (int i = 0; i < 4; ++i) {
        difference[i] = initial_rooms[i] - final_rooms[i];
    }

    return difference;
}

int main() {
    int A, B, C, D;
    cin >> A >> B >> C >> D;

    int X, Y;
    cin >> X >> Y;

    vector<int> result = room_optimization(A, B, C, D, X, Y);

    // Output the difference in the number of rooms for each type
    for (int i = 0; i < 4; ++i) {
        cout << result[i] << " ";
    }

    return 0;
}
